src/flux/store.js

```diff
@@ -234,21 +239,28 @@ class Store extends EventEmitter {
   };
 
   handleNewLog = (message) => {
-    const { data } = message;
+    const { data: log } = message;
 
-    const log = data;
-
-    log.formattedTimestamp = new Date(log.created * 1000).toLocaleString();
+    log.timestamp = new Date(log.created * 1000);
+    log.formattedTimestamp = log.timestamp.toLocaleString();
     log.idx = _store.logs.length;
 
     _store.logs.push(log);
     _store.processes[log.process] = log.name;
-    _store.logSources[log.name] = true;
-    _store.logLevels[log.levelname] = true;
 
-    if (CHART_LEVELS.includes(log.levelname)) {
-      _store.occurences.current[log.levelname]++;
-    }
+    if (_store.logSources[log.name]) _store.logSources[log.name]++;
+    else _store.logSources[log.name] = 1;
+
+    if (_store.logLevels[log.levelname]) _store.logLevels[log.levelname]++;
+    else _store.logLevels[log.levelname] = 1;
+
+    const unixTime = parseInt(log.timestamp / 1000);
+
+    if (!_store.logLevelOccurences[unixTime])
+      _store.logLevelOccurences[unixTime] = getInitialLevelOccurences();
+
+    _store.logLevelOccurences[unixTime].levels[log.levelname]++;
+    _store.logLevelOccurences[unixTime].lastLog = log.idx;
 
     this.emit("update-logs");
   };
```

Comment:
We should not multiply the timestamp with 1000 and later divide it by 1000 and parse it into an int.

Reply:
```javascript
const bucketSize = 0.5;

const result = logs.reduce(
  (acc, curr) => {
    const { created } = curr;
    const diff = created - acc.lastBucket;
    const isNextBucket = diff > bucketSize;
    if (isNextBucket) {
      acc.lastBucket = created;
      acc[created] = 0;
    }
    acc[acc.lastBucket]++;
    return acc;
  },
  { [logs[0].created]: 0, lastBucket: logs[0].created }
);
```

something like that could work but there's for sure better ways to do it.

In general:
rendering things is very expensive, depriving things from state is cheap, most of the times, and if it isn't, it's easier to memoize calculations starting from computing stuff from state.
keeping state, that it computed/derived from other state and has to be kept in sync with that state is a source of many bugs and causes high costs because it will require developers to invest a lot of time fixing bugs or onboarding, so they know about those things.